src/physics/em/AtomicRelaxationParams.cc

```diff
@@ -0,0 +1,250 @@
+ */
+void AtomicRelaxationParams::append_element(const ElementInput& inp)
+{
+    AtomicRelaxElement result;
+
+    // TODO: For an element Z with n shells of transition data, you can bound
+    // this worst case as n for radiative transitions and 2^n - 1 for
+    // non-radiative transitions if for a given vacancy the transitions always
+    // originate from the next subshell up. Physically this won't happen, so
+    // can we bound this tighter (maybe O(100) for non-radiative transitions,
+    // but that's still a lot)? Can we impose an energy cut below which we
+    // won't create secondaries in the relaxation cascade to reduce it further?
+    if (is_auger_enabled_)
+    {
+        result.max_secondary = std::pow(2, inp.shells.size()) - 1;
+    }
+    else
+    {
+        result.max_secondary = inp.shells.size();
+    }
+
+    // Copy subshell transition data
+    result.shells = this->extend_shells(inp);
+
+    // Add to host vector
+    host_elements_.push_back(result);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Process and store electron subshells to the internal list.
+ */
+Span<AtomicRelaxSubshell>
+AtomicRelaxationParams::extend_shells(const ElementInput& inp)
+{
+    CELER_EXPECT(host_shells_.size() + inp.shells.size()
+                 <= host_shells_.capacity());
+
+    // Allocate subshells
+    auto start = host_shells_.size();
+    host_shells_.resize(start + inp.shells.size());
+    Span<AtomicRelaxSubshell> result{host_shells_.data() + start,
+                                     inp.shells.size()};
+
+    // Create a mapping of subshell designator to index in the shells array
+    des_to_id_.clear();
+    for (SubshellId::value_type i : range(inp.shells.size()))
+    {
+        des_to_id_[inp.shells[i].designator] = SubshellId{i};
+    }
+    CELER_ASSERT(des_to_id_.size() == inp.shells.size());
+
+    for (auto i : range(inp.shells.size()))
+    {
+        // Store the radiative transitions
+        auto fluor = this->extend_transitions(inp.shells[i].fluor);
+
+        // Append the non-radiative transitions if Auger effect is enabled
+        if (is_auger_enabled_)
+        {
+            auto auger = this->extend_transitions(inp.shells[i].auger);
+            result[i].transitions = {fluor.begin(), auger.end()};
+        }
+        else
+        {
+            result[i].transitions = fluor;
+        }
+
+        // For a given subshell vacancy, EADL transition probabilities are
+        // normalized so that the sum over all radiative and non-radiative
+        // transitions is 1
+        real_type norm = std::accumulate(
+            result[i].transitions.begin(),
+            result[i].transitions.end(),
+            real_type(0),
+            [](real_type sum, const AtomicRelaxTransition& tr) {
+                return sum + tr.probability;
+            });
+        CELER_ASSERT(soft_near(1., norm, 1.e-5));
```

Comment:
Ah yes, will do. Also I see I messed this up with my last set of changes; the probabilities should sum to 1 over radiative and non-radiative transitions, which they won't here if Auger is disabled.. which means I should probably extend my test coverage. Why renormalize during import and not here, though? Won't it be the same?

Reply:
As far as our code goes, yes, but if we end up sharing our IO routines I think it would be good to guarantee normality as a postcondition. Good catch with the normalizing for no-auger...

Reply:
Ahh I see, thanks.